jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java

```diff
@@ -0,0 +1,693 @@
+
+        CacheEntry it = cachedOperations.get(i).get(j);
+        while (it != null) {
+            if (it.editOperation == null) {
+                result.addAll(editOperationLists.get(it.treeDistanceI).get(it.treeDistanceJ));
+            } else {
+                result.add(it.editOperation);
+            }
+            it = cachedOperations.get(it.cachePreviousPosI).get(it.cachePreviousPosJ);
+        }
+
+        return result;
+    }
+
+    /**
+     * Auxiliary class which for computes keyroot vertices,
+     * tree ordering and $l()$ function for a particular tree.
+     *
+     * <p>
+     * A keyroot of a tree is a vertex which has a left sibling.
+     * Ordering of a tree assings an integer index to every its vertex.
+     * Indices are assigned using post-order traversal. $l()$ function
+     * for every vertex in a tree returns ordering index of its leftmost child.
+     * For leaf vertex the function returns its own ordering index.
+     */
+    private class TreeOrdering {
+        /**
+         * Underlying tree of this ordering.
+         */
+        final Graph<V, E> tree;
+        /**
+         * Root vertex of {@code tree}.
+         */
+        final V treeRoot;
+
+        /**
+         * List of keyroots of {@code tree}.
+         */
+        List<Integer> keyroots;
+
+        /**
+         * List which at very position $i$ stores a vertex from {@code tree}
+         * which has ordering index equal to $i$.
+         */
+        List<V> indexToVertexList;
+        /**
+         * List which at every position $i$ stores value of $l()$ function
+         * for a vertex from {@code tree} whihc has ordering index equal to $i$.
+         */
+        List<Integer> indexToLValueList;
+        /**
+         * Ordering index to be assigned to the next traversed vertex.
+         */
+        int currentIndex;
+
+        /**
+         * Constructs an instance of the tree ordering for the given
+         * {@code graph} and {@code treeRoot}.
+         *
+         * @param tree     a tree
+         * @param treeRoot root vertex of {@code tree}
+         */
+        public TreeOrdering(Graph<V, E> tree, V treeRoot) {
+            this.tree = tree;
+            this.treeRoot = treeRoot;
+
+            int numberOfVertices = tree.vertexSet().size();
+            keyroots = new ArrayList<>();
+            indexToVertexList = new ArrayList<>(Collections.nCopies(numberOfVertices + 1, null));
+            indexToLValueList = new ArrayList<>(Collections.nCopies(numberOfVertices + 1, null));
+            currentIndex = 1;
+
+            computeKeyrootsAndMapping(treeRoot);
+        }
+
+        /**
+         * Runs post-order DFS on {@code tree} starting at {@code treeRoot}.
+         * Assigns consecutive integer index to every traversed vertex and
+         * computes keyroots for {@code tree}.
+         *
+         * @param treeRoot root vertex of {@code tree}
+         */
+        private void computeKeyrootsAndMapping(V treeRoot) {
```

Comment:
This procedure uses linearized recursion. I will leave the original version of the recursive procedure to make the reviewing process more convenient.

```java
private int computeKeyrootsAndMapping(V v, List<V> stack, boolean isKeyroot) {
    stack.add(v);
    
    V vParent = null;
    if (stack.size() > 1) {
        vParent = stack.get(stack.size() - 2);
    }
    
    boolean isKeyrootArg = false;
    int lValue = -1;
    for (V vChild : Graphs.successorListOf(tree, v)) {
        if (vParent == null || !vChild.equals(vParent)) {
            int lVChild = computeKeyrootsAndMapping(vChild, stack, isKeyrootArg);
            isKeyrootArg = true;
            if (lValue == -1) {
                lValue = lVChild;
            }
        }
    }
    
    if (lValue == -1) { // this is a leaf
        lValue = currentIndex;
    }
    
    if (isKeyroot) {
        keyroots.add(currentIndex);
    }
    indexToVertexMap.set(currentIndex, v);
    indexToLMap.set(currentIndex, lValue);
    ++currentIndex;
    stack.remove(stack.size() - 1);
    
    return lValue;
}
```